### PR TITLE
Preemption

### PR DESCRIPTION
**Pull request by jhjourdan**
Two changes:

1. As discussed in MPR#7669, use nanosleep as an implementation of yield for better preemption
2. Make Thread.delay an alias for Unix.sleepf. There is no longer any point for using select here. Moreover, Unix.sleepf supports (theoretically) nanosecond precision.

### CODE REVIEW

**Comment by gasche** [comment]
There are CI failures that need to be investigated. Interestingly, the `lib-threads/signal.ml` test failed on both amd64 and amd64+flambda testing machines. Does the added fairness make the signal heisen-failure reproducible? (cc @xavierleroy who may be delighted to have a fighting chance to understand this lingering testsuite bug)

I asked the CI to re-run one of them, just in case.

**Comment by jhjourdan** [comment]
The other failure is on Windows : the test I added for checking that preemption works fails.

It seems like Sleep(0) is not doing what it says. Should I disable the test on windows? How do I do that? Or does someone has an idea how to fix this on Windows?

**Comment by gasche** [comment]
Before you disable the Windows test, I think it would be nice to understand what `Sleep(0)` actually does.

On how to disable the test: see `tests/lib-unix/win-stat/Makefile` which does exactly the opposite (disable a test on non-Windows machines).

**Comment by jhjourdan** [comment]
And I am not able to reproduce the signal.ml failure....

File: otherlibs/systhreads/st_posix.h
Comment (xavierleroy) [inline otherlibs/systhreads/st_posix.h]: This should be `#ifdef __linux__` and it should be put above the two comment lines.  Initially at least we want to restrict the change of implementation to Linux, where the problem is known.  Later you can test MacOS and all the BSDs to see which is better, `nanosleep` or `sched_yield`.

File: testsuite/tests/lib-systhreads/testpreempt.ml
Comment (xavierleroy) [inline testsuite/tests/lib-systhreads/testpreempt.ml]: Pardon my French but this test is awful.  On a very fast machine (in the future possibly) the `long_computation` may well terminate before 0.2 s causing unexpected output.  On a slow machine or in bytecode the `long_computation` may take a long time and slow down the test suite uselessly.  Please use `Sys.time` to iterate a medium-size computation until 2 seconds are elapsed.

File: testsuite/tests/lib-systhreads/testpreempt.ml
Comment (xavierleroy) [inline testsuite/tests/lib-systhreads/testpreempt.ml]: "Interact", or better "Interaction", not "Interract".

**Review by xavierleroy — CHANGES_REQUESTED** [review]
Some comments below.  I'm really not convinced that the test should go in the test suite, because unfairness is not a bug, just a quality of implementation issue, and many scheduling accidents can cause it to fail, e.g.CI virtual machines that run on heavily loaded hardware.

File: testsuite/tests/lib-systhreads/testpreempt.ml
Comment (jhjourdan) [reply testsuite/tests/lib-systhreads/testpreempt.ml]: I changed the test. The reason I did not do that in the first place is that I wondered that the system calls in `Sys.time` would artificially trigger preemption. But since `Sys.time` does not release the master lock , there is no risk.

**Comment by jhjourdan** [comment]
I granted all @xavierleroy's wishes, except one: I left the test in the test suite (strengthened as you requested, though). The fact that this is an actual bug or not is arguable, but this can cause real bugs in applications containing UIs. So, if we want users to be able to rely on this better behavior, it better be tested.

**Comment by xavierleroy** [comment]
Running the current code through CI precheck gives:
image
Some of the Windows machines are offline right now (light blue dots), but the Windows configurations that were tested are OK.  Most of the Unix configurations fail (red dot) because of the dreaded `lib-threads/signal.ml` test.

**Comment by jhjourdan** [comment]
We still have to wait for the CI to complete, but I think I fixed the `signal.ml` test.

What was happening is that the runner of this test did not wait for the process to terminate, and the checker was actually ran before the process finished to write its output.

As for Windows: I find it strange that Inria CI succeeds while AppVeyor fails. Is there a way of looking in detail the results produced by the AppVeyor CI?

**Comment by xavierleroy** [comment]
"Precheck" in progress at Inria CI, stay tuned.

**Comment by xavierleroy** [comment]
CI results look very good indeed:
image
(Ignore the light blue / light red / light gray entries, these are configurations that were not tested because the CI is a bit clogged at the moment.)

So, the race condition was in the runner shell script, while I was desperately looking for it in the Caml code...  Well spotted indeed, and thanks a lot for solving this mystery!

File: otherlibs/systhreads/st_posix.h
Comment (xavierleroy) [inline otherlibs/systhreads/st_posix.h]: It is nicer to use `NULL` as second argument of `nanosleep` since it makes it obvious that you don't need the remaining time to be stored in `t`.

**Review by xavierleroy — APPROVED** [review]
Looks very good to me.  Could be merged as is but I can't help adding one more nitpick, see below.  Follow or ignore the suggestion, as you wish.

File: otherlibs/systhreads/st_posix.h
Comment (jhjourdan) [reply otherlibs/systhreads/st_posix.h]: Done.

**Comment by jhjourdan** [comment]
I still have no idea why appveyor is failing, but Inria's precheck isn't.

The only reason I can imagine is that the bug only occurs with msvc in 64 bits mode, which was not available with Inria's CI. @xavierleroy, do you think it would be possible to re-enable the ocaml-msvc-64 machine?

**Comment by shindere** [comment]
Jacques-Henri Jourdan (2017/12/20 13:35 +0000):
> The only reason I can imagine is that the bug only occurs with msvc in
> 64 bits mode, which was not available with Inria's CI. @xavierleroy,
> do you think it would be possible to re-enable the ocaml-msvc-64
> machine?

The machine seems up in Jenkins...

**Comment by xavierleroy** [comment]
The CI backlog cleared up overnight, so I could finish the test on ocaml-msvc-64.  It passes just fine.

I have no idea what's going on with Appveyor, and can't even see how to access the test log files.  But remember my prediction that weird scheduling by the OS can always cause `testpreempt.ml` to fail.  Could Appveyor have that kind of weird scheduling?

At any rate, I'd love to merge this PR now but cannot because of the risk of breaking all Appveyor testing.  So, could you please turn off the `testpreempt.ml` test, at least under Windows?  It's either this or the PR gets stuck until someone debugs the Appveyor behavior.

**Comment by dra27** [comment]
I'm just investigating the AppVeyor failure (sorry for not doing so earlier)

**Comment by dra27** [comment]
@xavierleroy, @shindere - you say it's working on the Inria machines. Can you remind me which version of Visual Studio they're running? I've just seen this failure on an Azure host running msvc64 on VS 2015. There are no "Interaction" lines the test output.

**Comment by xavierleroy** [comment]
Thanks @dra27 for looking into this.  Concerning the Visual Studio version, the ocaml-msvc-64 machine reports the following:
```
$ type cl
cl is /cygdrive/c/Program Files (x86)/Microsoft Visual Studio 12.0/VC/BIN/amd64/cl

$ cl -?
Microsoft (R) C/C++ Optimizing Compiler Version 18.00.40629 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```
Let me know if you need more information.

**Comment by dra27** [comment]
Actually, I think the question about Inria CI *might* be somewhat more subtle - how many CPUs do the test slaves have?

**Comment by xavierleroy** [comment]
The plot thickens :-)  ocaml-msvc-64 reports two cores, and its overall throughput is consistent with having two cores, but it's a virtual machine provided by a Cloudstack installation, so that's an approximation of the truth.  We run two builds&tests in parallel on this machine, one with flambda and the other without.  Each build&test is sequential.

**Comment by xavierleroy** [comment]
PS.  As far as I remember, the other Windows test machines (msvc, mingw, cygwin, 32 and 64) have the same configuration as ocaml-msvc-64: two-core Cloudstack virtual machines running two builds&test in parallel.

**Comment by dra27** [comment]
I think I'm seeing this fail on all 4 ports - though I didn't check that run very thoroughly (I will later this evening - I don't like being unsure of these things, in case you hadn't noticed!).

I was wondering if it might be to do with processor affinity - my reading suggests that `Sleep(0)` can't yield to a thread scheduled on another CPU. So if the main thread and the secondary thread are on different CPUs, then the main thread never yields. I haven't managed to test this in a way I'm truly satisfied with.

What does seem to work very reliably is changing the `Sleep(0)` to `Sleep(1)` in `st_thread_yield` in `st_win32.h`. That does make a certain amount of sense, since it must force the thread to be suspended. Not sure what the performance implication of doing that every 50ms is, though?

**Comment by dra27** [comment]
OK, there was something up with my msvc32 environment, but the test fails consistently in mingw32, mingw64 and msvc64 on an 8 vCPU Windows Server 2016 instance.

**Comment by jhjourdan** [comment]
Thanks, @dra27, for your help investigating this.

> What does seem to work very reliably is changing the Sleep(0) to Sleep(1) in st_thread_yield in st_win32.h. That does make a certain amount of sense, since it must force the thread to be suspended. Not sure what the performance implication of doing that every 50ms is, though?

In theory, that would mean a 2% performance loss in throughput, which is quite a bit. A possible way to avoid this issue would be to yield every 100ms (or more).

**Comment by xavierleroy** [comment]
I am very afraid that `Sleep(1)` could end up sleeping one full clock tick.  Again, we won't compromise througput in the name of fairness.  

An alternative to be considered is `SwitchToThread`.

**Comment by dra27** [comment]
@xavierleroy - SwitchToThread documents the CPU affinity (IIRC) and I already experimented with it yesterday. It behaves the same as `Sleep(0)`.

I have to say that this behaviour doesn't surprise me on Windows - I'd either expect to have to raise the scheduling priority of one thread, or manually insert yields into the main thread.

**Comment by dra27** [comment]
I haven't tested this with the Cygwin ports, because they seemed to be broken (possibly my setup), but I think you are right that this test should be disabled at least on native Windows with a clear reference as to why.

File: testsuite/tests/lib-systhreads/testpreempt.precheck
Comment (dra27) [inline testsuite/tests/lib-systhreads/testpreempt.precheck]: `"$UNIX_OR_WIN32 != "win32"` would be better.

**Review by dra27 — CHANGES_REQUESTED** [review]
I'll test the Cygwin ports properly just before this gets merged.

**Comment by jhjourdan** [comment]
Ok, I've just disabled the test on Windows. We just have to wait for the CI to succeed, and then this PR should be ready to merge.

**Comment by dra27** [comment]
You've been caught out by my hastily typed comment - there's a double-quote missing! CI won't catch a Cygwin fault - give me a few minutes!

**Comment by dra27** [comment]
Cygwin is affected too - so you also want to test for `"$SYSTEM" != "cygwin"` (NB both Cygwin and Cygwin64 set SYSTEM=cygwin, unlike the native ports which differentiate 32/64 bit).

**Comment by jhjourdan** [comment]
Ok, I'm testing for:

```
test "$OS" != "Windows_NT"
```

This should work on all Windows ports, right?

**Comment by dra27** [comment]
It's a user-controllable variable and as of the latest insider build you can see that in WSL (interesting aside as to whether this test passes on Ubuntu running on Windows!). Ok today, maybe not tomorrow - but then what ever is ok tomorrow! :)

**Review by dra27 — APPROVED** [review]

**Comment by xavierleroy** [comment]
It's merged!  Thanks to all who participated.

**Comment by xavierleroy** [comment]
FYI: we're not done with the signal.ml test...  Now, sometimes it blocks forever.  Commit c08532f makes sure that the main thread terminates eventually, even if no signal is received, but the blocking behavior persists.  Either the OCaml process enters an infinite loop after receiving the signal (?), or the `wait` in the shell script misses the termination of the OCaml process (??).

**Comment by xavierleroy** [comment]
And sometimes it is testpreempt.ml that times out... I'm starting to suspect Jenkins CI itself rather than a problem with OCaml's test suite.  Stay tuned.

**Comment by jhjourdan** [comment]
Would it be possible that this is a deadlock in the signal handling mechanism? This could be fixed by #1128.

**Comment by xavierleroy** [comment]
I was able to reproduce the deadlock and attach gdb.  Indeed this is related to #1128 in that the Caml code for the signal handler is run while one of the threads holds the lock for the stdout channel, so the signal handling code waits forever for this lock.  Until #1128 is ready, I'll fix the test by not doing any I/O from the signal handler.

**Pull request merged**
